File: app/src/main/java/com/ocram/qichwadic/framework/dao/FavoriteDao.java
package com.ocram.qichwadic.framework.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.ocram.qichwadic.domain.model.Favorite;

import java.util.List;

import io.reactivex.Flowable;

@Dao
public interface FavoriteDao {

    @Query("SELECT * from favorite ORDER BY word, id")
    Flowable<List<Favorite>> getFavorites();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addFavorite(Favorite favorite);

    @Delete
    void removeFavorite(Favorite favorite);

    @Query("DELETE from favorite")
    void clearFavorites();
}

File: app/src/main/java/com/ocram/qichwadic/core/data/local/dao/FavoriteDao.kt
package com.ocram.qichwadic.core.data.local.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query

import com.ocram.qichwadic.core.data.model.FavoriteEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface FavoriteDao {

    @Query("SELECT * from favorite ORDER BY word, id LIMIT :limit OFFSET :offset")
    fun getFavorites(limit: Int = 20, offset: Int = 0): Flow<List<FavoriteEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addFavorites(vararg favorite: FavoriteEntity): List<Long>

    @Delete
    suspend fun removeFavorite(favorite: FavoriteEntity): Int

    @Query("DELETE from favorite")
    suspend fun clearFavorites(): Int
}
